File: TEMPLATE.py
#!/usr/bin/python

import re

class TemplateItem:
	def __init__(self, parent, htmlFile, html=None, section="ROOT", uniqID="ROOT"):
		if parent is None:
			with open(htmlFile, 'r') as myFile:
				data = myFile.read()
		else:
			data = html		
	
		self.Parent = parent
		self.Html = data
		self.Section = section
		self.UniqID = uniqID
		
		self.ListTemplateItem = {}
		self.HtmlTemplate = {}

		self.initAllAvailableSection()
	

	def initAllAvailableSection(self):
		
		strPattern = r'<!--[ ]*BEGIN[ ]*' + self.Section + r'\.([a-zA-Z0-9]+)[ ]*-->[ ]*(.*?)[ ]*<!--[ ]*END[ ]*' + self.Section + r'\.([a-zA-Z0-9]+)[ ]*-->'
		pattern = re.compile(strPattern, re.MULTILINE|re.DOTALL)

		for m in re.finditer(pattern, self.Html):
			self.ListTemplateItem[m.group(1)] = []
			self.HtmlTemplate[m.group(1)] = m.group(2)
		
		self.HashVariable = {}		

	def AddVar(self, variable, valeur):
		self.HashVariable[variable] = valeur

	def initNewSection(self, initSection, nom):
		htmlSection = self.HtmlTemplate[initSection]

		if htmlSection is None:
			return None
		
		template = TemplateItem(self, None, htmlSection, initSection, nom)
		
		self.ListTemplateItem[initSection].append(template)

		return template

	def GetHtmlParsed(self):
		htmlTemplate = self.Html
		
		for key in self.HashVariable.keys():
			htmlTemplate = htmlTemplate.replace('<VARSUB>' + key + '</VARSUB>', self.HashVariable[key])
				
		return htmlTemplate

	def HaveChild(self):
		for key in self.ListTemplateItem.keys():
			if len(self.ListTemplateItem[key]) > 0 :
				return true

		return false


	def Parse(self):
		result = self.GetHtmlParsed()

		for key in self.ListTemplateItem.keys():
			info = ""
			print "SECTION " + key
			listTemplate = self.ListTemplateItem[key]


			for template in listTemplate:
				if template.HaveChild :
					info = template.Parse()
				else:
					return self.GetHtmlParsed()
		
			result = result.replace(self.HtmlTemplate[key], info)
			
			strPattern = r'<!--[ ]*(BEGIN|END)[ ]*' + self.Section + r'\.' + key + r'[ ]*-->'
			
			pattern = re.compile(strPattern, re.MULTILINE|re.DOTALL)
			
			result = pattern.sub('', result)
			
		return result
		


x = TemplateItem(None, "TEMPLATE.HTML")

x2 = x.initNewSection('TEST', '')

x2.AddVar("TATA", "GOOD JOB!")

print x.Parse()

File: Template.java
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class Template
{
	private Template pParent;
	private String pHtml;
	private String pSection;
	private String pUniqId;

	private HashMap<String, List<Template>> pListTemplateItem = new HashMap<String, List<Template>>();
	private HashMap<String, String> pHtmlTemplate = new HashMap<String, String>();  
	private HashMap<String, String> pHashVariable = new HashMap<String, String>();

	private static String PATTERN = "<!--[ ]*BEGIN[ ]*__SECTION__.([a-zA-Z0-9]+)[ ]*-->[ ]*(.*?)[ ]*<!--[ ]*END[ ]*__SECTION__.([a-zA-Z0-9]+)[ ]*-->";

	public Template(String fileTemplateHtml) throws FileNotFoundException, IOException
	{
		BufferedReader br = new BufferedReader(new FileReader(fileTemplateHtml));
		String line = br.readLine();
		pHtml = "";		

		while(line != null)
		{
			pHtml += line;
			line = br.readLine();
		}
		
		pSection = "ROOT";
		pUniqId = "ROOT";

		initAllAvailableSection();
	}

	public Template(Template parent, String html, String section, String uniqId)
	{
		pParent = parent;
		pHtml = html;
		pSection = section;
		pUniqId = uniqId;

		initAllAvailableSection();		
	}

	private void initAllAvailableSection()
	{
		Pattern p = Pattern.compile(PATTERN.replaceAll("__SECTION__", pSection));
		
		Matcher m = p.matcher(pHtml);

		while(m.find())
		{
			pListTemplateItem.put(m.group(1), new ArrayList<Template>());		
			pHtmlTemplate.put(m.group(1), m.group(2));
		}
	}	

	public void addVar(String variable, String valeur)
	{
		pHashVariable.put(variable, valeur);
	}

	public Template initNewSection(String initSection, String nom)
	{
		String htmlSection = pHtmlTemplate.get(initSection);

		if(htmlSection == null)
			return null;

		Template t = new Template(this, htmlSection, initSection, nom);

		pListTemplateItem.get(initSection).add(t);

		return t;		
	}

	public String getHtmlParsed()
	{
		String htmlTemplate = pHtml;
		
		for(String key : pHashVariable.keySet())
		{
			htmlTemplate = htmlTemplate.replace("<VARSUB>" + key + "</VARSUB>", pHashVariable.get(key));		
		}

		return htmlTemplate;
	}

	public boolean haveChild()
	{
		for(String key : pListTemplateItem.keySet())
		{
			if(pListTemplateItem.get(key).size() > 0 )
				return true;			
		}

		return false;
	}

	public String parse()
	{
		String result = getHtmlParsed();
		
		for(String key : pListTemplateItem.keySet())
		{
			String info = "";
			
			for(Template t : pListTemplateItem.get(key))
			{
				if(haveChild())
					info += t.parse();
				else
					return result;

			}

			result = result.replace(pHtmlTemplate.get(key), info);
				
			result = result.replaceAll("<!--[ ]*(BEGIN|END)[ ]*" + pSection + "." + key + "[ ]*-->", "");
			//result = result.replaceAll("<!--[ ]*BEGIN[ ]*" + pSection + "." + key + "[ ]*-->", "");
			//result = result.replaceAll("<!--[ ]*END[ ]*" + pSection + "." + key + "[ ]*-->", "");
		}

		return result;	
	}

	public static void main(String args[]) throws FileNotFoundException, IOException
	{
		Template root = new Template("TEMPLATE.html");
	
		Template t2 = root.initNewSection("TEST", "");

		t2.addVar("TATA", "GOOD JOB!");

		Template th = root.initNewSection("TH", "");
		th.addVar("TITLE", "T");			

		Template rows = root.initNewSection("ROWS", "");
		Template td = rows.initNewSection("TD", "");
		td.addVar("DATA", "connard!");

		//root.parse();
		System.out.println(root.parse());

	} 
}
